### PR TITLE
Vision flipping mitigation from RoboCup 2019

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description
Changes from RoboCup 2019

### Testing Done
Validated on the field

### Resolved Issues
None

### Length Justification
None

### CODE REVIEW

File: src/thunderbots/software/ai/main.cpp [hunk 1]
```diff
@@ -32,14 +32,30 @@ void worldUpdateCallback(const thunderbots_msgs::World::ConstPtr &msg)
...
-    world.updateEnemyTeamState(new_world.enemyTeam());
-    world.updateFriendlyTeamState(new_world.friendlyTeam());
-    world.updateTimestamp(new_world.getMostRecentTimestamp());
-    RefboxGameState new_game_state =
-        Util::ROSMessages::createGameStateFromROSMessage(world_msg.refbox_data.command);
-    world.updateRefboxGameState(new_game_state);
+    // During RoboCup 2019, we saw weird cases where vision frames wold randomly appear
+    // mirrored. This filter checks if our goalie appears in the enemy defense area to
+    // detect mirroring, and ignores the data if mirroring is detected
+    if (!Util::DynamicParameters::AI::vision_flipping_filter.value() ||
```
Comment (garethellis0) [inline src/thunderbots/software/ai/main.cpp]: Could we change `vision_flipping_filter` to `vision_flipping_filter_enabled`?

File: src/thunderbots/software/ai/main.cpp [hunk 2]
```diff
@@ -32,14 +32,30 @@ void worldUpdateCallback(const thunderbots_msgs::World::ConstPtr &msg)
...
+        !(new_world.friendlyTeam().goalie() &&
+          new_world.field().enemyDefenseArea().containsPoint(
+              new_world.friendlyTeam().goalie()->position())))
```
Comment (garethellis0) [inline src/thunderbots/software/ai/main.cpp]: Could we pull out the second bit of this into a `bool` like `friendly_goalie_in_enemy_defense_area`? I had some trouble reading through and understanding this initially.

File: src/thunderbots/software/util/parameter/config/ai_control.yaml [hunk 3]
```diff
@@ -1,4 +1,8 @@
 AI:
+  vision_flipping_filter:
+    type: "bool"
+    default: true
+    description: Ignores frames if our goalie appears in the opponent crease
```
Comment (garethellis0) [inline src/thunderbots/software/util/parameter/config/ai_control.yaml]:
``` suggestion
    description: Ignores frames if our goalie appears in the opponent defense area
```
(just to match up a bit better with what we call it in code)

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Sorry, meant to request changes

File: src/thunderbots/software/ai/main.cpp (hunk 1 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/main.cpp]: done

File: src/thunderbots/software/ai/main.cpp (hunk 2 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/main.cpp]: done

File: src/thunderbots/software/ai/main.cpp [hunk 4]
```diff
@@ -32,14 +32,32 @@ void worldUpdateCallback(const thunderbots_msgs::World::ConstPtr &msg)
...
-    world.updateRefboxGameState(new_game_state);
+    // During RoboCup 2019, we saw weird cases where vision frames wold randomly appear
+    // mirrored. This filter checks if our goalie appears in the enemy defense area to
+    // detect mirroring, and ignores the data if mirroring is detected
+    bool friendly_goalie_in_enemy_defense_area =
+        new_world.friendlyTeam().goalie() &&
+        new_world.field().enemyDefenseArea().containsPoint(
+            new_world.friendlyTeam().goalie()->position());
+    if (!Util::DynamicParameters::AI::vision_flipping_filter_enabled.value() ||
+        !friendly_goalie_in_enemy_defense_area)
```
Comment (garethellis0) [inline src/thunderbots/software/ai/main.cpp]: nit; not blocking, but could we change this to `Util::DynamicParameters::AI::vision_flipping_filter_enabled.value() && friendly_goalie_in_enemy_defense_area`? Logically equivalent and easier to follow

**Comment by MathewMacDougall** [comment]
@garethellis0 	ready for re-review

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Pull request merged**